nautobot/docs/development/jobs/job-patterns.md

```diff
@@ -0,0 +1,537 @@
+
+This page provides a catalog of **common patterns** used when building Jobs, especially those that go beyond the basics of the `run()` method. These examples are designed to be **practical, modular, and reusable**-you can pick and choose the parts that make sense for your own Job logic.
+
+Because Jobs are highly flexible, many common patterns tend to emerge depending on what the Job is trying to accomplish - like importing data, validating configuration, logging results, or returning output to users.  
+
+Not every Job needs every feature shown below. Use this page as a reference: copy/paste and customize only the patterns you need.
+
+## Common Job Patterns
+
+- [Accessing User and Job Context](#accessing-user-and-job-context)
+- [Logging Patterns](#logging-patterns)
+    - Logging levels
+    - Structured log context with `extra`
+    - Skipping database logging
+    - Markdown formatting and sanitization
+- [Job Flow Control](#job-flow-control)
+    - Guard clauses with validation
+    - Success and failure conditions
+    - Marking a Job as failed
+- [Working with Files](#working-with-files)
+    - File output with `create_file()`
+    - Uploading and parsing files (`FileVar`)
+    - Reading static YAML or JSON files
+- [Reference Implementations](#reference-implementations)
+    - Device validation
+    - Creating objects for a planned location
+    - Everything demo from the Example App
+
+## Job Execution Context and Logging Patterns
+
+When writing Jobs, it's helpful to access information about the current execution - like which user ran the Job or which JobResult is being updated. It's equally important to log progress or results in a way that's visible in both the Nautobot UI and REST API.
+
+This section covers how to use the built-in logger, including structured log messages, context-aware logging, and optional enhancements like skipping database logging or using Markdown formatting.
+
+### Accessing User and Job Context
+
+The user who initiated the Job and the associated `JobResult` are available via the `self.user` and `self.job_result` properties of the Job class.
+
+This information is useful for:
+
+- **Personalization** — customizing behavior or output based on the user
+- **Auditing** — logging who ran what, and when
+- **Conditional logic** — restricting actions to admins or specific users
+- **Debugging** — correlating job logs with user sessions or API activity
+
+Here's a simple example that logs which user launched the Job, along with its unique result ID:
+
+<!-- pyml disable-num-lines 10 proper-names -->
+!!! example
+    ```py
+    username = self.user.username
+    job_result_id = self.job_result.id
+    self.logger.info("Job %s initiated by user %s is running.", job_result_id, username)
+    ```
+
+You can access other standard Django user attributes like `self.user.email`, `self.user.is_staff`, or use `self.user.has_perm(...)` for permission checks.
+
+### Logging Patterns
```

Comment:
What do you think about `Job Logging` being its own page? I didn't expect to see it under common patterns, so I'm thinking it's just a bit hidden here. And important enough to maybe warrant its own section.

Reply:
I think that is totally fair. I was struggling on which category to put it under whether under structure or patterns, but I think like you said it should need its own section.

Reply:
I think that is a good idea, yeah it never really fit well in the patterns area, but I didn't want too many small sections. I will go ahead and add it now